REGRESSION(262585@main): Potential null dereference of cached images in MemoryCache::destroyDecodedDataForAllImages()

shallawa opened the pull request:
#### 1fc9a2418bc9fb920b9f7c49b6e9e3963701c3de
<pre>
REGRESSION(262585@main): Potential null dereference of cached images in MemoryCache::destroyDecodedDataForAllImages()
<a href="https://bugs.webkit.org/show_bug.cgi?id=257982">https://bugs.webkit.org/show_bug.cgi?id=257982</a>
rdar://110339514

Reviewed by Chris Dumez.

262585@main added a call to MemoryCache::destroyDecodedDataForAllImages() when
the cleanup timer fires. This function may try to release the decoded data of an
already deleted CachedImage.

MemoryCache stores the list of cached resources as WeakPtrs. So we need to null
check these WeakPtrs in forEachResource() before using them.

* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::forEachResource):

Canonical link: <a href="https://commits.webkit.org/265093@main">https://commits.webkit.org/265093@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f186b0c81fcf199a055a7661b59b0a2f74c1d538

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9804 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10052 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11456 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9813 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12036 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10002 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12466 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9958 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10769 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8074 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8900 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/11876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9172 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9048 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9543 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8711 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2348 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12937 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9317 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/cache/MemoryCache.cpp:
```diff
@@ -222,8 +222,11 @@ void MemoryCache::pruneLiveResources(bool shouldDestroyDecodedDataForAllLiveReso
     for (auto& unprotectedLRUList : m_allResources) {
-        for (auto& resource : copyToVector(*unprotectedLRUList))
+        for (auto& resource : copyToVector(*unprotectedLRUList)) {
+            if (!resource)
```
Given this is only one line, I would have done:
```
if (resource)
     function(resource);
```

but ok.

cdumez reviewed (approved):
LGTM

shallawa committed:
REGRESSION(262585@main): Potential null dereference of cached images in MemoryCache::destroyDecodedDataForAllImages()
https://bugs.webkit.org/show_bug.cgi?id=257982
rdar://110339514

Reviewed by Chris Dumez.

262585@main added a call to MemoryCache::destroyDecodedDataForAllImages() when
the cleanup timer fires. This function may try to release the decoded data of an
already deleted CachedImage.

MemoryCache stores the list of cached resources as WeakPtrs. So we need to null
check these WeakPtrs in forEachResource() before using them.

* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::forEachResource):

Canonical link: https://commits.webkit.org/265093@main
